juju/model.py

```diff
@@ -168,26 +213,34 @@ def __getitem__(self, key):
 
 # We mix in MutableMapping here to get some convenience implementations, but whether it's actually
 # mutable or not is controlled by the flag.
-class RelationUnitData(LazyMapping, MutableMapping):
-    def __init__(self, relation, unit, is_mutable, backend):
+class RelationEntityData(LazyMapping, MutableMapping):
+    def __init__(self, relation, entity, is_mutable, backend):
         self.relation = relation
-        self.unit = unit
+        self.entity = entity
         self._is_mutable = is_mutable
         self._backend = backend
 
     def _load(self):
         try:
-            return self._backend.relation_get(self.relation.id, self.unit.name)
+            return self._backend.relation_get(self.relation.id, self.entity.name, app=False)
         except RelationNotFound:
             # Dead relations tell no tales (and have no data).
             return {}
 
     def __setitem__(self, key, value):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
+            # This might happen due to a mid-hook leadership change besides trying to set data in remote data bags.
```

Comment:
Can that really happen?  If so, ugh; I don't think any charm would handle that well and would just blow up.